src/vmm/src/logger/metrics.rs

```diff
@@ -692,7 +692,84 @@ impl SignalMetrics {
     }
 }
 
-/// Metrics specific to VCPUs' mode of functioning.
+/// Provides efficient way to record LatencyAggregateMetrics
+#[derive(Debug)]
+pub struct LatencyMetricsRecorder<'a> {
+    start_time: u64,
+    metric: &'a LatencyAggregateMetrics,
+}
+
+impl<'a> LatencyMetricsRecorder<'a> {
+    /// Const default construction.
+    fn new(metric: &'a LatencyAggregateMetrics) -> Self {
+        Self {
+            start_time: utils::time::get_time_us(utils::time::ClockType::Monotonic),
+            metric,
+        }
+    }
+}
+impl<'a> Drop for LatencyMetricsRecorder<'a> {
+    /// records aggregate (min/max/sum) for the given metric
+    /// This captures delta between self.start_time and current time
+    /// and updates min/max/sum metrics.
+    ///  self.start_time is recorded in new() and metrics are updated in drop
+    fn drop(&mut self) {
```

Comment:
It seems like the assumption here is that `drop()` will be called once we exit the function that created `LatencyMetricsRecorder` variable. Is there any type of optimization that might cause the `drop` method to be executed at a different time? Maybe @roypat or @JonathanWoollett-Light know something about it?

Reply:
drop will be called when the lifetime of the variable end i.e. when the brackets {} end.
I tried to add some comments on this [here](https://github.com/firecracker-microvm/firecracker/pull/4346/files#diff-e13e5029c4b049ea5100ce89b82d716a8f1d71241aa29344b57c793441509722R751)

Reply:
The drop scopes are specified in https://doc.rust-lang.org/reference/destructors.html, so I'd say anything that differs from that would be a rustc bug